Use SymEngine in sympy/physics/mechanics

isuruf opened the pull request:
When environment variable `USE_SYMENGINE` is set to 1, underlying symbolic engine for sympy.physics is switched to SymEngine

Timings for `n_link_pendulum_on_cart(n)`

| n | SymEngine + SymPy | SymPy only | Speedup |
| --- | --- | --- | --- |
| 10 | 0.17 s | 5.58 s | 32.8x |
| 15 | 0.44 s | 22.07 s | 50.1x |
| 20 | 0.95 s | 45.59 s | 47.9x |
| 30 | 3.11 s | 162.80 s | 52.3x |
| 40 | 7.02 s | 427.16 s | 60.8x |
| 50 | 13.95 s | 915.83 s | 65.6x |
| 60 | 23.16 s | 1596.37 s | 68.9x |

Needs latest [symengine/symengine](https://github.com/symengine/symengine) master and [symengine/symengine.py](https://github.com/symengine/symengine.py) to run.

cc @certik, @moorepants

isuruf committed:
Use SymEngine

moorepants commented:
This is awesome. So simple now that you all have these objects available in SymEngine.

This seems fine to merge as an undocumented feature for now. There will of course need to be a larger discussion on how to officially support this for SymPy.

What does this mean for the outputs of n_link_pendulum. Now that it outputs symengine objects, will they be able to be used in the code generation steps form SymPy's code? Or any other SymPy functions?

isuruf commented:
If all SymPy functions `sympify` their inputs, then there shouldn't be any problem as it would convert SymEngine objects to SymPy, although for speed, this is not desirable.
I'll have to look at code generation code to see if SymEngine objects can be used directly without converting to SymPy. Is there any particular code generation method you are interested in?

certik commented:
Wow, you are awesome. Great job. This looks very good to me.

Sent from my mobile phone.
On Apr 12, 2016 7:25 PM, "Isuru Fernando" <notifications@github.com> wrote:

> When environment variable USE_SYMENGINE is set to 1, underlying symbolic
> engine for sympy.physics and sympy.vector is switched to SymEngine
>
> Timings for n_link_pendulum_on_cart(n)
> n SymEngine + SymPy SymPy only
> 10 0.43 s 5.58 s
> 15 1.61 s 22.07 s
> 20 3.20 s 45.59 s
> 30 9.32 s 162.80 s
> 40 22.02 s 427.16 s
> 50 41.69 s 915.83 s
> 60 83.31 s .....................
>
> Needs latest symengine/symengine <https://github.com/symengine/symengine>
> master and symengine/symengine.py#58
> <https://github.com/symengine/symengine.py/pull/58> to run.
>
> cc @certik <https://github.com/certik>, @moorepants
> <https://github.com/moorepants>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/sympy/sympy/pull/11002
> Commit Summary
>
>    - Use SymEngine
>
> File Changes
>
>    - *M* sympy/physics/mechanics/functions.py
>    <https://github.com/sympy/sympy/pull/11002/files#diff-0> (6)
>    - *M* sympy/physics/mechanics/kane.py
>    <https://github.com/sympy/sympy/pull/11002/files#diff-1> (11)
>    - *M* sympy/physics/mechanics/particle.py
>    <https://github.com/sympy/sympy/pull/11002/files#diff-2> (7)
>    - *M* sympy/physics/vector/frame.py
>    <https://github.com/sympy/sympy/pull/11002/files#diff-3> (12)
>    - *M* sympy/physics/vector/functions.py
>    <https://github.com/sympy/sympy/pull/11002/files#diff-4> (15)
>    - *M* sympy/physics/vector/vector.py
>    <https://github.com/sympy/sympy/pull/11002/files#diff-5> (13)
>
> Patch Links:
>
>    - https://github.com/sympy/sympy/pull/11002.patch
>    - https://github.com/sympy/sympy/pull/11002.diff
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/sympy/sympy/pull/11002>
>

certik commented:
@isuruf  I updated your table with speedups. How is it possible that the speedup itself gets larger with `n`? I thought symengine delivers just a constant speedup, but this is clearly not the case here.

moorepants commented:
Is it simply the algorithm we use in KanesMethod?

certik commented:
@moorepants my understanding is that sympy and symengine is used in exactly the same way, so no matter what algorithm is used, the speedup should be constant. I think it probably suggests that SymPy is not using optimal algorithms compared to symengine.

isuruf commented:
@certik, from what I've observed, SymPy uses redundant calculations in the core and the code hasn't been touched in years.
For example, let `p/q` be a Rational with `p` and `q` its numerator and denominator.
You would expect `p/q + 0` to be of constant time complexity. This is true in SymEngine, but it is not in SymPy. In SymPy, this would calculate `(p+0*q)/q` and triggers a gcd call. I tried to fix this in #10539, but things fail when I do this. See #10928 as well.

I'm not sure what causes the speedup increase here though, but I'm not surprised it does increase.

moorepants commented:
Isuru. Maybe it would be worth making some sympy benchmarks in the benchmarks repo. We can ten find which commits introduced the slow downs and try reverting or improving them. I have a computer setup in my office to run them and output to here: http://www.moorepants.info/misc/sympy-asv/

isuruf commented:
@moorepants, I looked at code generation, specifically lambdify. Lambdify needs a tweak in PyDy to work with SymEngine and it is faster by about a factor of 2.5x at ODE integration steps. (This is normal for same code written in Python vs C++)

moorepants commented:
Having "fast to compile/slow(er) to execute" code gen is beneficial for verifying that models numerically correct. That's mainly what lambdify is used for. But for any large problem or desired fast execution, you want the "slower to compile/fastest to execute". These essentially correspond to the lambdify and cython backends are in PyDy.

isuruf commented:
Note that the timings are on a machine with 2 physical cores and openmp enabled. Otherwise the speedup is reduced by about 1.5 times for n=50 case.

isuruf commented:
Updated the table with the timings from latest master.
`msubs` was written in C++, after refactoring `subs` into a visitor (similar to SymPy's printer). It requires only a few line in C++ in https://github.com/symengine/symengine/blob/master/symengine/subs.h#L242-L255 but gives a very good speedup.

@certik, I added `sympy.core.backend` from which you can import types like Matrix, Integer etc. and would import symengine or sympy types according to env var. This is similar to what you mentioned in the mailing list.

This PR is ready for review now.

isuruf commented:
ping

moorepants commented:
This looks fine as is. I don't see any issue with merging as it doesn't affect anything if SymEngine is not installed. But in the long run we need some regression tests that run on travis with SymEngine installed. We also need something in the docs that specifies what version of SymEngine this depends on and think about how to handle version dependencies.

isuruf commented:
There's a conda package now for symengine. I'll use that for running tests on travis.

certik commented:
Awesome, let's install it on Travis to test this. Then we can use it at more places in SymPy.

@isuruf thanks for all the work that you are doing. It's all tremendously useful.

isuruf commented:
@asmeurer, how do you want testing to be done?
New job or reuse an existing job? If existing job, which one?
In allow_failures or not?

isuruf committed:
Use SymEngine in more places

isuruf commented:
@certik, there are some tests that uses `x(t).diff()` where the differentiation variable is not explicitly given.
Derivative method would inspect the free_symbols and use the variable from free_symbols if it contains only one variable. Otherwise an error is thrown.
Do we want this behaviour in symengine.py? Or should the tests be changed?

certik commented:
I would change the test if at all possible. The diff() way seems to be
meant for interactive use. I wouldn't implement it in symengine, unless we
have to.

Sent from my mobile phone.
On Jun 20, 2016 12:54 AM, "Isuru Fernando" <notifications@github.com> wrote:

> @certik <https://github.com/certik>, there are some tests that uses
> x(t).diff() where the differentiation variable is not explicitly given.
> Derivative method would inspect the free_symbols and use the variable from
> free_symbols if it contains only one variable. Otherwise an error is thrown.
> Do we want this behaviour in symengine.py? Or should the tests be changed?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sympy/sympy/pull/11002#issuecomment-227063403>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AABQWL9ep-z4Ycy2QLP-0gtZvyskm1Giks5qNjkigaJpZM4IF-EF>
> .
>

isuruf commented:
For dynamic symbols it does make sense, though.

isuruf committed:
SymEngine fixes

isuruf committed:
Test SymEngine in travis

isuruf commented:
I've added tests. Let me know what you think

isuruf committed:
Merge branch 'master' of github.com:sympy/sympy into pydy

isuruf committed:
Fix test failures

isuruf committed:
sympify matrices before equal

certik commented on sympy/core/backend.py:
```diff
@@ -0,0 +1,27 @@
+        sin, cos, tan, cot, csc, sec, asin, acos, atan, acot, acsc, asec,
+        sinh, cosh, tanh, coth, asinh, acosh, atanh, acoth, lambdify,
+        symarray, diff, zeros, eye, diag, ones, zeros, expand,
+        Function, symbols, var)
```
These two lists are not equivalent. For example you import Add/Mul from sympy but not symengine. Shouldn't the two lists be exactly equivalent?

certik commented on sympy/physics/vector/dyadic.py:
```diff
@@ -1,4 +1,8 @@
+if USE_SYMENGINE:
+    from sympy.core.backend import Matrix
+else:
+    from sympy import ImmutableMatrix as Matrix
```
Is `sympy.ImmutableMatrix` faster than `sympy.Matrix`? While in symengine we only have (mutable) Matrix, so that's what we need to use here. And that's why you cannot just use `from sympy.core.backend import Matrix`, but have to add the switch here to get a faster code if `USE_SYMENGINE==False`?

isuruf replied:
I don't have an idea why ImmutableMatrix is used, but I'm going to implement a ImmutableMatrix type in symengine.py (not symengine). This is a temporary fix. symengine/symengine.py#76

I didn't want to change anything in SymPy, that's why sympy version wasn't changed to a mutable matrix.

certik commented:
This looks very good overall, left some minor comments. Overall +1 from me.

@asmeurer can you please review this one? This is the start of using symengine in sympy, so we better get it right. It seems the solution that @isuruf implemented is very clean --- it introduces the `sympy.core.backend` module, which uses either sympy or symengine, depending on the `USE_SYMENGINE` environment variable.

SymPy code can either import directly from sympy or sympy.core, then SymPy is used. Or it imports from `sympy.core.backend`, then either sympy or symengine will be used.

We don't know yet whether we can make the symengine native Python API 100% compatible with sympy (because of the different internal representation, for speed reasons we want to reserve a right to differ in some places from SymPy, in the future), but if we have to differ, we'll simply write a SymPy compatible wrapper and expose it in the `sympy.core.backend`. So the code in sympy or user code can pretty much use `sympy.core.backend` as a drop in replacement. As you can see from this PR, a pretty non-trivial part of SymPy just works, and gets very impressive speedups.

This approach is also non-invasive. We'll only use `sympy.core.backend` in sympy where all tests pass and we make sure things work. Everywhere else, as well as the default `from sympy import *` is just SymPy, as it was before, so user code won't be affected. The only way a user code can get affected by this change is if the user code inspects the internals of sympy mechanics *and* if the user enables `USE_SYMENGINE` which is off by default. In such a case, we'll either fix the bug, or tell the user to disable `USE_SYMENGINE` until his or our code is fixed.

asmeurer commented on sympy/physics/mechanics/functions.py:
```diff
@@ -532,7 +539,7 @@ def _recurser(expr, sub_dict):
 def _fraction_decomp(expr):
     """Return num, den such that expr = num/den"""
-    if not isinstance(expr, Mul):
+    if not isinstance(expr, MulClass):
```
What is the difference between Mul and MulClass?

certik replied:
Yeah, looking into `sympy.core.backend` it looks like `Mul` and `MulClass` are exactly the same thing, but in SymPy and SymEngine. @isuruf will have to answer this one.

asmeurer replied:
Ditto for `tan` and `Tan`. This worries me. If the SymEngine wrappers don't use the same names for functions and their classes like SymPy does, then compatibility lines like this will be a PITA to support in SymPy.

isuruf replied:
In SymPy, `Add(x, x)` works and returns a `Mul`. Also `x + y` is an instance of `Add`.

But for SymEngine, we can't have both. Hence `Add` is the class and `add` will sum all the terms given.

certik replied:
For this PR, this should be fixed so that `sympy.core.backend` provides exactly the same names as sympy does, e.g. `from symengine import Tan as tan`.

In general, we will hit the problem that some small things might be different (perhaps some method, or such). Then I think we need to use this solution:

https://github.com/symengine/symengine/wiki/SymPy-core-upgrade-to-SymEngine

certik replied:
@isuruf we can have both in symengine.py, we just need to implement this in the wrappers (you are right that in C++ we can't do that). I don't now if we should, since things will get inevitably more complex.

We can however write a compatible layer on top of the native symengine.py, or use the approach described in https://github.com/symengine/symengine/wiki/SymPy-core-upgrade-to-SymEngine.

asmeurer replied:
So I take it C++ doesn't let you dispatch to a different type in the constructor like Python does. 

It should be possible to use a single name `Add` in the Python wrapper. You just need a class `Add` whose `__new__` calls your `add` function and whose `__instancecheck__` returns `True` for subclasses of the actual wrapper `Add` class. This will break if any code in SymPy uses `type(x) is Add` instead of `isinstance(x, Add)`, but [it shouldn't be doing that anyway](https://github.com/sympy/sympy/issues/6751).

isuruf replied:
`tan` is the function and `Tan` is the type in SymEngine. Is there a way to have both be referred using `tan`?

asmeurer replied:
If `Tan` is the Python class, then `Tan.__new__(x)` should return `tan(x)`.

asmeurer replied:
Actually, I guess you have to use a metaclass. See https://docs.python.org/3.5/reference/datamodel.html#customizing-instance-and-subclass-checks.

certik replied:
@asmeurer yes, we can implement that in the wrappers, as you described.

certik replied:
But the problem is that it will complicate the wrappers. I am not convinced we should go that way. Because there will be more subtle differences.

I would rather write a compatibility layer for sympy, so that we have truly equivalent interface, on top of the symengine.py.

asmeurer replied:
That makes no difference to me. Should the compatibility layer be part of SymEngine or part of SymPy?

certik replied:
It could be part of either library.

It technically makes a huge difference to me. But it seems we both agree that we need a 100% (or close to 100%) compatibility layer, somewhere. Whether built on top of symengine.py, or whether symengine.py itself (i.e. the Cython wrapper) would be 100% compatible, those are various technical solutions to the same, ultimate problem.

We should discuss what the compatibility layer should have. E.g. what I have in mind is that if you do things like

a = sin(x)
b = a.diff(x)

Then "b" will have exactly the same methods, whether you call it with sympy or symengine backend.

I am worried about speed implications caused by this compatibility layer. I can see pros and cons of each technical solution that I have in mind.

Right now I lean towards just making our best effort to clean this PR, to make it reasonable clean to merge, and then investigate long term solutions.

asmeurer replied:
Yes, it should all be done in a compatibility layer. Basically, for this to work, there should be no code anywhere in SymPy that does

```py
if HAS_SYMENGINE:
    ...
else:
    ...
```

That also means that the SymEngine core has to have all the same methods as SymPy (even if some of them just dispatch to SymPy Python code). For instance, if the SymEngine core doesn't have the `diff` method, then any code that takes a derivative in SymPy will have to use the above `if HAS_SYMENGINE` to do something different for the SymPy core and the SymEngine core, OR, nowhere in SymPy can use the diff method, only the function, meaning that the diff method is basically dropped from the API (and what if a user wants to use `a.diff(x)` in their code?).  

It really needs to "just work". Instead of failing with AttributeError or whatever when something isn't implemented, it should fallback to SymPy Python code. 

Regarding the speed, I would suggest just writing the simplest thing, and seeing how it affects the speed. For instance, I'm assuming you are worried that having a metaclass will slow things down. If it does, maybe there are ways to speed it up (like writing the compatibility layer in C).

certik replied:
@asmeurer I think the compatibility layer has nothing to do with C or C++, it's just how the Python wrappers are arranged. I agree that the
```
if HAS_SYMENGINE:
    ...
else:
    ...
```
should only be in the compatibility layer, and once you import `sin`, it needs to be 100% SymPy compatible (no matter how it's implemented underneath). The method described in https://github.com/symengine/symengine/wiki/SymPy-core-upgrade-to-SymEngine will just work and always fallback to sympy. It's just that it will constantly convert between sympy and symengine for each operation, i.e. if you call a.diff(x), where "a" and "x" are sympy objects, it will convert them to symengine, do the diff (in C++) and convert back to sympy. You can of course instead return the symengine Python expression from a.diff(x), but then this symengine Python object must have all the methods and behave like a sympy object. Then of course, `.args` must return sympy compatible objects and so on.

One motivation behind https://github.com/symengine/symengine/wiki/SymPy-core-upgrade-to-SymEngine is that it's slow, but ultimately, it will eventually call symengine for all operations that symengine can do. So you can literally remove most of sympy.core (if we decide to only use symengine --- which I think we shouldn't, see below), 
and then you can remove the conversions to and from sympy. Only at this point things will get fast, since no more conversions are done. One can then speed things up by calling into symengine.py directly, bypassing the compatibility layer.

The problem is that this assume we'll remove sympy.core. I wouldn't, I would always keep it around, I think it's good to have a pure Python implementation. So I think we need a different solution, one that allows us to incrementally use symengine and be fast from the very beginning, yet compatible with sympy.core.

The advantage is that the symengine/symengine repo has no Python, so we make sure that one is fast. The Python wrappers are slow just because we use Python. So if we make it a bit more slower by being compatible with sympy, I think we actually don't care much about that, because if there is something that needs to be fast, we simply implement it in C++, make it fast, and then just wrap it into Python.

One solution would be to have two wrappers ---- one set of wrappers that are as fast as one can get in Cython, but not compatible with sympy. The other wrappers would be 100% compatible with sympy, but slower. But it's a lot of work to maintain two sets of wrappers.

And so perhaps we can just make the wrappers themselves 100% compatible with sympy.

Ok, maybe I found a solution. I was always imagining the wrappers to have sympy objects, convert to symengine, call C++, and convert back to sympy. However, we should go the other way around --- each class like `sin` would implement all the methods of the sympy's sin, and each method would accept either sympy or symengine objects, convert to sympy, call sympy, and convert to symengine and return the result. So `sin(x).diff(x)` would return symengine's `cos(x)`. That way, if the `sin(x)` was already a symengine object, then no conversion is taking place and things are fast. At the same time, this `cos(x)` that it returns is a symengine object, that will however implement all sympy's methods, and so for the rest of sympy, it would be a drop in replacement.

And so that way, I think we can actually make symengine a drop in replacement for sympy objects, step by step. And things will still be fast, as long as the user keeps using symengine objects.

But here is another problem. Say complex numbers --- for speed reasons, `1+I` is a Complex class in C++. It is an `Add(1, I)` in SymPy. I don't think there is a way to make this 100% compatible. In other words, if `a.diff(x)` returns `1+I`, it will be a Complex class. SymPy doesn't have such a class. So things will break. I don't have a solution to this problem.

The other way, i.e. keeping things as sympy objects by default, as described here:

https://github.com/symengine/symengine/wiki/SymPy-core-upgrade-to-SymEngine

doesn't have this problem, since we never return a symengine object, we always convert it first to sympy, so a Complex class `1+I` becomes SymPy's `Add(1,I)`. So that's why I wrote this up in the wiki, as I know this will deliver. But this approach will necessarily be slower, since there are these conversions back and forth. Only at the very end, after all of sympy uses the compatibility layer and never sympy.core directly, only then one can remove the conversions back and forth, and things become fast. But that's a lot of work, and in the meantime, things might even become slower when you turn on symengine, since for simple things like `x+x` you have to convert back and forth all the time.

This PR is a middle ground. It doesn't try to be 100% compatible, it doesn't do any conversions and returns symengine objects. Thus it is fast.

Anyway, I don't know the best way forward.

certik replied:
I think the way forward is still to return symengine objects and try to be 100% compatible by calling into sympy for things that are not yet implemented, and convert back to symengine before returning.

This has the "1+I is a Complex class" problem. However, this can be fixed by simply converting to sympy, whenever this causes a problem in user's code. I expect there actually won't be that many places in sympy that require I to be ImaginaryUnit, as opposed to Complex. We can design a compatibility function, say, "is_imaginary_unit()", which would be implemented in the compatibility layer, and in there, if sympy is used, it would simply use `is_instance(ImaginaryUnit)`, and if symengine is used, it would query the `Complex` class. Then one just has to replace all occurrence of `is_instance(ImaginaryUnit)` in sympy with `is_imaginary_unit()`.

Now imagine some algorithm in sympy requires `1+I` to be an `Add`. Then in this very code, we simply call sympify() to convert into sympy from symengine. And we are done. Then, if we care about speed here, we can implement this whole algorithm in C++.

So I think this way we can fix all such issues. It's sort of like the Python 2 vs 3 workarounds --- Python 3 removes some classes, does some stuff differently, and so we introduce compatibility layer, and we must use it throughout. 

I think this can work. What do you think?

asmeurer replied:
Couldn't Complex pretend to be an Add, for the purposes of its SymPy methods (like .args)? I suppose an Add with a complex would also need to pretend that it is a part of it (like `Add(1, I, x).args` should give `(1, I, x)`, not `(Complex(1+I), x)`).  

That's the sort of thing that the wrapper would have to do, fake the SymPy API everywhere. If the SymPy API is bad, say, because it forces a slowdown, we can discuss if it should be changed.

certik replied:
It can't, the Complex class would have to literally convert to an Add. And
of course, symengine Add can't hold 1+I, you would get an assert error in
debug mode. So what you are proposing is my proposal on the wiki, i.e.
convert everything to sympy before returning. And that would indeed work,
but would be slow until we port everything.

So my alternative proposal anove is better I think.

Sent from my mobile phone.
On Jul 1, 2016 4:33 PM, "Aaron Meurer" <notifications@github.com> wrote:

> In sympy/physics/mechanics/functions.py
> <https://github.com/sympy/sympy/pull/11002#discussion_r69362963>:
>
> > @@ -532,7 +539,7 @@ def _recurser(expr, sub_dict):
> >
> >  def _fraction_decomp(expr):
> >      """Return num, den such that expr = num/den"""
> > -    if not isinstance(expr, Mul):
> > +    if not isinstance(expr, MulClass):
>
> Couldn't Complex pretend to be an Add, for the purposes of its SymPy
> methods (like .args)? I suppose an Add with a complex would also need to
> pretend that it is a part of it (like Add(1, I, x).args should give (1,
> I, x), not (Complex(1+I), x)).
>
> That's the sort of thing that the wrapper would have to do, fake the SymPy
> API everywhere. If the SymPy API is bad, say, because it forces a slowdown,
> we can discuss if it should be changed.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sympy/sympy/pull/11002/files/4c8b2d16e1820a104c81993c5905763e6698844c#r69362963>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AABQWLIShxJ2IKoPA_iLjQmPhZDCP0d-ks5qRZWigaJpZM4IF-EF>
> .
>

isuruf replied:
I'm sorry, I meant in symengine.py. Cython works a little different than regular Python classes. I'l check whether `__new__` thing works. If so, then `Add(x + x)` can work.

isuruf replied:
In Cython `__new__` doesn't work.
```Cython
Error compiling Cython file:
------------------------------------------------------------
...
        import sage.all as sage
        return self.real_part()._sage_() + sage.I * self.imaginary_part()._sage_()

cdef class Add(Basic):

    def __new__(cls, *args):
   ^
------------------------------------------------------------

symengine_wrapper.pyx:964:4: __new__ method of extension type will change semantics in a future version of Pyrex and Cython. Use __cinit__ instead.
```

Also when using ``__cinit__`` you cannot return a different type
```Cython
Error compiling Cython file:
------------------------------------------------------------
...
        return self.real_part()._sage_() + sage.I * self.imaginary_part()._sage_()

cdef class Add(Basic):

    def __cinit__(cls, *args):
        return add(args)
                 ^
------------------------------------------------------------

symengine_wrapper.pyx:965:18: Return with value in void function
```

isuruf replied:
> It should be possible to use a single name Add in the Python wrapper. You just need a class Add whose __new__ calls your add function and whose __instancecheck__ returns True for subclasses of the actual wrapper Add class. This will break if any code in SymPy uses type(x) is Add instead of isinstance(x, Add), but it shouldn't be doing that anyway.

I didn't read this correctly. What you are saying is possible. `Add` class have to be a pure Python class and it would work.

asmeurer commented:
The mechanism that this is done by (sympy.core.backend) doesn't feel right to me. So as more and more things are able to support symengine, more and more imports throughout SymPy will be changed to use `from sympy.core.backend import ...` instead of `from sympy import ...`? Also, most of the stuff in sympy.core.backend isn't part of sympy.core. 

Unfortunately, I don't have a better solution to suggest.

asmeurer commented on sympy/core/backend.py:
```diff
@@ -0,0 +1,27 @@
+        FunctionSymbol as AppliedUndef, Derivative as DerivativeClass,
+        Sin as SinClass, Cos as CosClass, Tan as TanClass)
+else:
+    import sympy as sm
```
Is this intentional? The one place where you imported `sm` below you imported this module as `sm`, not `sm` from this module. I personally wouldn't have this here. `sm` isn't even the shorthand I'd recommend for SymPy. I always tell people to use `sym`, but it's far from standardized.

certik replied:
Some people also use `sp`. I think `sym` is good.

asmeurer replied:
`sp` is bad because that's the standard shorthand for scipy.

certik replied:
Ah that's right. Maybe I saw that for scipy.

isuruf replied:
In `mechanics/models.py` there was a statement like `import sympy as sm`. That's why I initially had this, but these 2 statements in `backend.py` can be removed, because you can just do
`from sympy.core import backend as sm` instead of `import sympy as sm`

certik commented:
@asmeurer I have a better solution, described in details here:

https://github.com/symengine/symengine/wiki/SymPy-core-upgrade-to-SymEngine

Though I think we can move to this latter solution even after this PR is merged.

isuruf committed:
Use symengine/sympy_compat

certik commented:
Create `sympy.mechanics.backend` and put there the `USE_SYMENGINE` if statements. Remove `USE_SYMENGINE` from all the PR, except the `sympy.core.backend` and `sympy.mechanics.backend`.

Otherwise it looks good.

isuruf committed:
Remove unnecessary changes

isuruf committed:
Merge branch 'master' into pydy

isuruf commented on sympy/core/backend.py:
```diff
@@ -0,0 +1,23 @@
+        sin, cos, tan, cot, csc, sec, asin, acos, atan, acot, acsc, asec,
+        sinh, cosh, tanh, coth, asinh, acosh, atanh, acoth,
+        lambdify, symarray, diff, zeros, eye, diag, ones, zeros,
+        expand, Function, symbols, var, Add, Mul, Derivative)
```
@certik, with the new changes, notice that horrible `AddClass` workarounds are not there anymore.

certik replied:
I see. That's good. Is performance affected by this change?

isuruf replied:
Didn't notice any slowdown

asmeurer replied:
Just a heads up, you will likely run into this issue as this gets integrated with more of SymPy (things that use the wrong pattern like `type(x) is Add` instead of `isinstance(x, Add)` will not work) https://github.com/sympy/sympy/issues/6751

isuruf committed:
Fix whitespace

isuruf commented:
Tests pass now

certik commented on .travis.yml:
```diff
@@ -306,6 +310,12 @@ before_install:
+
+  - if [[ "${TEST_SYMENGINE}" == "true" ]]; then
+        conda config --add channels conda-forge;
+        conda config --add channels symengine;
```
How is the `symengine` channel updated? Should we only use `conda-forge`, and perhaps keep the `symengine` channel commented out, for quick debugging of PRs, but the master should depend on the conda-forge version?

certik replied:
Also, why don't we release a symengine, so that sympy master always depends on a released version of symengine (i.e. with a git tag).

certik replied:
Also, we need to specify explicitly somehow, which version of symengine is required. I think conda allows that, doesn't it? So if we release symengine, can we use the git tag to reference a conda package?

isuruf replied:
SymEngine channel is updated by this, https://github.com/symengine/python-symengine-feedstock/tree/symengine-tokens. I change the commit number and the bump the build number to update it.

We can use a git tag of symengine and build and use that version from `conda-forge` channel. I've had to make lots of changes to python wrappers and therefore I'm not sure we want to make a tag everytime I do a change to the python wrappers for sympy compatibility.

We can always do a symengine release for each sympy release.

certik replied:
If not a tag, can we at least specify a git hash?

Sent from my mobile phone.
On Jul 24, 2016 9:18 AM, "Isuru Fernando" <notifications@github.com> wrote:

> In .travis.yml
> <https://github.com/sympy/sympy/pull/11002#discussion_r71990328>:
>
> > @@ -306,6 +310,12 @@ before_install:
> >          # this assumes TEST_THEANO runs in the same build as TEST_MATPLOTLIB
> >          conda install theano;
> >      fi
> > +
> > +  - if [[ "${TEST_SYMENGINE}" == "true" ]]; then
> > +        conda config --add channels conda-forge;
> > +        conda config --add channels symengine;
>
> SymEngine channel is updated by this,
> https://github.com/symengine/python-symengine-feedstock/tree/symengine-tokens.
> I change the commit number and the bump the build number to update it.
>
> We can use a git tag of symengine and build and use that version from
> conda-forge channel. I've had to make lots of changes to python wrappers
> and therefore I'm not sure we want to make a tag everytime I do a change to
> the python wrappers for sympy compatibility.
>
> We can always do a symengine release for each sympy release.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sympy/sympy/pull/11002/files/ad9ae61edeac795e1bc7a1d4e7ab7d010f42ded4#r71990328>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABQWNB8MJUMaylvq8qCzbJUJaCwZyeKks5qY4I2gaJpZM4IF-EF>
> .
>

certik commented:
Otherwise +1.

isuruf committed:
Pin symengine version

isuruf commented on .travis.yml:
```diff
@@ -306,6 +310,12 @@ before_install:
+  - if [[ "${TEST_SYMENGINE}" == "true" ]]; then
+        conda config --add channels conda-forge;
+        conda config --add channels symengine;
+        conda install python-symengine=0.1.0.dev0+97c5a21
```
@certik, pinned version `0.1.0.dev0+97c5a21`

certik replied:
Perfect, thanks! +1 to merge from me.

@asmeurer, would you mind reviewing it once more?

Sent from my mobile phone.
On Jul 24, 2016 11:13 PM, "Isuru Fernando" <notifications@github.com> wrote:

> In .travis.yml
> <https://github.com/sympy/sympy/pull/11002#discussion_r72009526>:
>
> > @@ -306,6 +310,12 @@ before_install:
> >          # this assumes TEST_THEANO runs in the same build as TEST_MATPLOTLIB
> >          conda install theano;
> >      fi
> > +
> > +  - if [[ "${TEST_SYMENGINE}" == "true" ]]; then
> > +        conda config --add channels conda-forge;
> > +        conda config --add channels symengine;
> > +        conda install python-symengine=0.1.0.dev0+97c5a21
>
> @certik <https://github.com/certik>, pinned version 0.1.0.dev0+97c5a21
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sympy/sympy/pull/11002/files/d909aaf58e3ec60ba6acdd704489774df200b693#r72009526>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABQWJSNtuYZTophGfq813mU8vpDU6wlks5qZEYNgaJpZM4IF-EF>
> .
>

asmeurer commented:
Looks fine for now. We may eventually want a more elegant way to do the cross-imports, but we can figure that out later.

bjodah commented:
Looks good to me +1

isuruf committed:
Merge remote-tracking branch 'upstream/master' into pydy

isuruf committed:
Use SymEngine in new mechanics/system

isuruf committed:
Use symengine=0.2.0

isuruf commented:
Can somebody mark this for 1.1 milestone or better yet review? I'd like to get this in before the next release.

isuruf commented:
Tests pass now

isuruf commented:
ping

asmeurer commented:
You can't add milestones?

isuruf commented:
No. I don't have access.

certik commented:
@isuruf you should have access now.

certik commented on sympy/physics/vector/functions.py:
```diff
@@ -609,5 +608,5 @@ def dynamicsymbols(names, level=0):
 
 
-dynamicsymbols._t = Symbol('t')
+dynamicsymbols._t = symbols('t')
```
This is not needed, is it? You can still import `Symbol` from `sympy.core.backend`, cannot you?

certik reviewed (approved):
I think this looks good enough to merge as it is. We can further improve it in master.

certik commented:
@isuruf feel free to go ahead and merge this.

asmeurer commented:
I see, you didn't have push access. I was worried that maybe only admins are allowed to add milestones, which wouldn't be good. 

When this is merged, please update the release notes.

isuruf commented:
Thanks for the reviews.

moorepants commented:
Thanks for this work. Sorry I haven't been able to review. First of the school year is sucking all my time.
